File: integration_tests/elixir-postgrex/test/integration_test.exs
alias Scythe.Queries

database_url =
  System.get_env("DATABASE_URL", "postgres://scythe:scythe@localhost:5432/scythe_test")

uri = URI.parse(database_url)
[username, password] = String.split(uri.userinfo, ":")
database = String.trim_leading(uri.path, "/")

{:ok, conn} =
  Postgrex.start_link(
    hostname: uri.host,
    port: uri.port || 5432,
    username: username,
    password: password,
    database: database
  )

# Clean slate
Postgrex.query!(conn, "DROP TABLE IF EXISTS user_tags CASCADE", [])
Postgrex.query!(conn, "DROP TABLE IF EXISTS tags CASCADE", [])
Postgrex.query!(conn, "DROP TABLE IF EXISTS orders CASCADE", [])
Postgrex.query!(conn, "DROP TABLE IF EXISTS users CASCADE", [])
Postgrex.query!(conn, "DROP TYPE IF EXISTS user_status CASCADE", [])
Postgrex.query!(conn, "DROP TYPE IF EXISTS user_address CASCADE", [])

schema_sql = File.read!(Path.join([__DIR__, "..", "..", "sql", "pg/schema.sql"]))
schema_sql
|> String.split(";")
|> Enum.map(&String.trim/1)
|> Enum.filter(&(&1 != ""))
|> Enum.each(fn stmt -> Postgrex.query!(conn, stmt, []) end)

exit_code = 0

assert = fn condition, test_name, detail ->
  unless condition do
    IO.puts(:stderr, "FAIL: #{test_name}: #{detail}")
    Process.put(:exit_code, 1)
  end
end

Process.put(:exit_code, 0)

# Test: CreateUser
{:ok, user} = Queries.create_user(conn, "Alice", "alice@example.com", "active")
assert.(user.name == "Alice", "CreateUser", "expected name Alice, got #{user.name}")
assert.(user.email == "alice@example.com", "CreateUser", "expected email alice@example.com")
assert.(user.status == "active", "CreateUser", "expected status active, got #{user.status}")
user_id = user.id
IO.puts("PASS: CreateUser")

# Test: GetUserById
{:ok, fetched} = Queries.get_user_by_id(conn, user_id)
assert.(fetched.id == user_id, "GetUserById", "expected id #{user_id}")
assert.(fetched.name == "Alice", "GetUserById", "expected name Alice")
assert.(fetched.email == "alice@example.com", "GetUserById", "expected email alice@example.com")
IO.puts("PASS: GetUserById")

# Test: ListActiveUsers
{:ok, active_users} = Queries.list_active_users(conn, "active")
assert.(length(active_users) > 0, "ListActiveUsers", "should have at least one user")
first = List.first(active_users)
assert.(first.name == "Alice", "ListActiveUsers", "first user should be Alice")
IO.puts("PASS: ListActiveUsers")

# Test: CreateOrder
{:ok, order} = Queries.create_order(conn, user_id, Decimal.new("99.95"), "first order")
assert.(order.user_id == user_id, "CreateOrder", "expected user_id #{user_id}")
assert.(Decimal.equal?(order.total, Decimal.new("99.95")), "CreateOrder", "expected total 99.95, got #{order.total}")
assert.(order.notes == "first order", "CreateOrder", "expected notes 'first order'")
IO.puts("PASS: CreateOrder")

# Test: GetOrdersByUser
{:ok, orders} = Queries.get_orders_by_user(conn, user_id)
assert.(length(orders) == 1, "GetOrdersByUser", "expected 1 order, got #{length(orders)}")
first_order = List.first(orders)
assert.(Decimal.equal?(first_order.total, Decimal.new("99.95")), "GetOrdersByUser", "expected total 99.95")
IO.puts("PASS: GetOrdersByUser")
# Test: GetUserProfile (board #197/#204) -- a nullable enum and a nullable
# composite column, each observed both present and as SQL NULL, plus a
# composite field containing a double quote and a comma. Postgrex decodes
# a composite into a native positional tuple, not text, so this exercises
# UserAddress.from_tuple rather than a text parser -- it still catches a
# regression that stopped decoding the composite at all (board #204).
present_sql =
  "INSERT INTO users (name, email, status, secondary_status, address) " <>
    "VALUES ('Carol', 'carol@example.com', 'active', 'inactive', " <>
    "ROW('1 Main St', 'Springfield', '12345')) RETURNING id"
{:ok, %{rows: [[present_id]]}} = Postgrex.query(conn, present_sql, [])

absent_sql =
  "INSERT INTO users (name, email, status, secondary_status, address) " <>
    "VALUES ('Dave', 'dave@example.com', 'active', NULL, NULL) RETURNING id"
{:ok, %{rows: [[absent_id]]}} = Postgrex.query(conn, absent_sql, [])

quoted_sql =
  "INSERT INTO users (name, email, status, secondary_status, address) " <>
    "VALUES ('Eve', 'eve@example.com', 'active', 'inactive', " <>
    "ROW('12 \"Main\", Apt 3', 'Berlin', '10115')) RETURNING id"
{:ok, %{rows: [[quoted_id]]}} = Postgrex.query(conn, quoted_sql, [])

{:ok, profile} = Queries.get_user_profile(conn, present_id)
assert.(profile.secondary_status == "inactive", "GetUserProfile", "expected secondary_status inactive, got #{profile.secondary_status}")
assert.(profile.address != nil, "GetUserProfile", "expected address to be present")
assert.(profile.address.street == "1 Main St", "GetUserProfile", "expected address.street '1 Main St', got #{profile.address.street}")
assert.(profile.address.city == "Springfield", "GetUserProfile", "expected address.city 'Springfield', got #{profile.address.city}")
assert.(profile.address.zip == "12345", "GetUserProfile", "expected address.zip '12345', got #{profile.address.zip}")

{:ok, null_profile} = Queries.get_user_profile(conn, absent_id)
assert.(null_profile.secondary_status == nil, "GetUserProfile", "expected secondary_status nil, got #{inspect(null_profile.secondary_status)}")
assert.(null_profile.address == nil, "GetUserProfile", "expected address nil, got #{inspect(null_profile.address)}")

{:ok, quoted_profile} = Queries.get_user_profile(conn, quoted_id)
assert.(quoted_profile.address != nil, "GetUserProfile", "expected quoted address to be present")
assert.(quoted_profile.address.street == "12 \"Main\", Apt 3", "GetUserProfile", "expected address.street '12 \"Main\", Apt 3', got #{quoted_profile.address.street}")
assert.(quoted_profile.address.city == "Berlin", "GetUserProfile", "expected address.city 'Berlin', got #{quoted_profile.address.city}")
assert.(quoted_profile.address.zip == "10115", "GetUserProfile", "expected address.zip '10115', got #{quoted_profile.address.zip}")

:ok = Queries.delete_user(conn, present_id)
:ok = Queries.delete_user(conn, absent_id)
:ok = Queries.delete_user(conn, quoted_id)
IO.puts("PASS: GetUserProfile")

# Test: DeleteUser (delete orders first due to FK)
{:ok, deleted_orders} = Queries.delete_orders_by_user(conn, user_id)
assert.(deleted_orders == 1, "DeleteUser", "expected 1 deleted order, got #{deleted_orders}")
:ok = Queries.delete_user(conn, user_id)
result = Queries.get_user_by_id(conn, user_id)
assert.(result == {:error, :not_found}, "DeleteUser", "user should not exist after deletion")
IO.puts("PASS: DeleteUser")

final_exit_code = Process.get(:exit_code, 0)

if final_exit_code == 0 do
  IO.puts("ALL TESTS PASSED")
end

System.halt(final_exit_code)

File: integration_tests/kotlin-exposed/src/main/kotlin/IntegrationTest.kt
import generated.*
import java.math.BigDecimal
import java.nio.file.Path
import kotlin.io.path.readText
import kotlin.system.exitProcess
import org.jetbrains.exposed.sql.Database
import org.jetbrains.exposed.sql.statements.StatementType
import org.jetbrains.exposed.sql.transactions.transaction

// Exposed is a DSL, not a driver, and that changes the harness shape: a
// generated query function takes no Connection. It opens its own
// `transaction { }` against the ambient Database that `Database.connect`
// installs, so this file connects once in main() and then calls the query
// functions with query arguments only -- unlike every kotlin-jdbc branch
// below, which threads `conn` through every test.

var passed = 0
var failed = 0

fun pass(name: String) {
    println("PASS: $name")
    passed++
}

fun fail(name: String, message: String) {
    println("FAIL: $name - $message")
    failed++
}

fun fail(name: String, e: Exception) {
    println("FAIL: $name - ${e.message}")
    failed++
}

var createdUserId = 0
var bobId = 0

fun main() {
    val databaseUrl = System.getenv("DATABASE_URL")
    if (databaseUrl.isNullOrEmpty()) {
        System.err.println("DATABASE_URL environment variable is required")
        exitProcess(1)
    }

    val uri = java.net.URI(databaseUrl)
    val userInfo = uri.userInfo?.split(":") ?: listOf("", "")
    val user = userInfo[0]
    val password = if (userInfo.size > 1) userInfo[1] else ""
    val jdbcUrl = "jdbc:postgresql://${uri.host}:${uri.port}${uri.path}"

    Database.connect(jdbcUrl, driver = "org.postgresql.Driver", user = user, password = password)

    runMigration()

    testCreateUser()
    testGetUserById()
    testUpdateUserEmail()
    testCreateOrder()
    testGetOrdersByUser()
    testGetOrderTotal()
    testListActiveUsers()
    testGetUserOrders()
    testCountUsersByStatus()
    testGetUserWithTags()
    testSearchUsers()
    testGetUserProfileNullable()
    testDeleteOrdersByUser()
    testDeleteUser()

    println()
    println("Results: $passed passed, $failed failed")
    if (failed > 0) {
        exitProcess(1)
    }
    println("ALL TESTS PASSED")
}

// The schema file holds several statements; Exposed's `exec` sends one at a
// time, so it is split the same way every other harness splits it. The
// fixture is comment-free and single-quote-free by the guard in
// tools/integration-test-generator/tests/coverage_completeness.rs, so a plain
// split on ';' is sound here.
fun runMigration() {
    val schemaPath = Path.of(System.getProperty("user.dir"))
        .resolve("../sql/pg/schema.sql")
        .normalize()
    val schema = schemaPath.readText()

    transaction {
        for (stmt in "DROP TABLE IF EXISTS user_tags CASCADE;DROP TABLE IF EXISTS tags CASCADE;DROP TABLE IF EXISTS orders CASCADE;DROP TABLE IF EXISTS users CASCADE;DROP TYPE IF EXISTS user_status CASCADE;DROP TYPE IF EXISTS user_address CASCADE".split(";")) {
            exec(stmt)
        }
        for (stmt in schema.split(";")) {
            if (stmt.isNotBlank()) {
                exec(stmt)
            }
        }
    }
}

fun testCreateUser() {
    val name = "CreateUser"
    try {
        val user = createUser("Alice", "alice@example.com", UserStatus.ACTIVE)
        if (user.name != "Alice") {
            fail(name, "expected name Alice, got ${user.name}")
            return
        }
        if (user.email != "alice@example.com") {
            fail(name, "expected email alice@example.com, got ${user.email}")
            return
        }
        if (user.status != UserStatus.ACTIVE) {
            fail(name, "expected status ACTIVE, got ${user.status}")
            return
        }
        createdUserId = user.id
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testGetUserById() {
    val name = "GetUserById"
    try {
        val user = getUserById(createdUserId)
        if (user.id != createdUserId) {
            fail(name, "expected id $createdUserId, got ${user.id}")
            return
        }
        if (user.name != "Alice") {
            fail(name, "expected name Alice, got ${user.name}")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testUpdateUserEmail() {
    val name = "UpdateUserEmail"
    try {
        updateUserEmail("alice-updated@example.com", createdUserId)
        val user = getUserById(createdUserId)
        if (user.email != "alice-updated@example.com") {
            fail(name, "expected updated email, got ${user.email}")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testCreateOrder() {
    val name = "CreateOrder"
    try {
        val order = createOrder(createdUserId, BigDecimal("99.99"), "Test order")
        if (order.user_id != createdUserId) {
            fail(name, "expected user_id $createdUserId, got ${order.user_id}")
            return
        }
        if (order.total.compareTo(BigDecimal("99.99")) != 0) {
            fail(name, "expected total 99.99, got ${order.total}")
            return
        }
        if (order.notes != "Test order") {
            fail(name, "expected notes 'Test order', got ${order.notes}")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testGetOrdersByUser() {
    val name = "GetOrdersByUser"
    try {
        val orders = getOrdersByUser(createdUserId)
        if (orders.size != 1) {
            fail(name, "expected 1 order, got ${orders.size}")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testGetOrderTotal() {
    val name = "GetOrderTotal"
    try {
        val result = getOrderTotal(createdUserId)
        val sum = result.total_sum
        if (sum == null) {
            fail(name, "returned null total_sum")
            return
        }
        if (sum.compareTo(BigDecimal("99.99")) != 0) {
            fail(name, "expected total_sum 99.99, got $sum")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

// The enum round-trip is the point of this one: the parameter goes out as the
// enum's SQL spelling and the rows come back decoded through fromValue, so a
// binding that sent the Kotlin enum object instead would fail here rather
// than somewhere downstream.
fun testListActiveUsers() {
    val name = "ListActiveUsers"
    try {
        val users = listActiveUsers(UserStatus.ACTIVE)
        if (users.isEmpty()) {
            fail(name, "expected at least 1 active user")
            return
        }
        if (users.none { it.name == "Alice" }) {
            fail(name, "expected Alice among active users")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testGetUserOrders() {
    val name = "GetUserOrders"
    try {
        val bob = createUser("Bob", "bob@example.com", UserStatus.ACTIVE)
        bobId = bob.id
        val rows = getUserOrders(UserStatus.ACTIVE)
        // Bob has no orders, so his LEFT JOIN row must carry a null total --
        // the nullable-side column the join widens.
        val bobRow = rows.firstOrNull { it.id == bobId }
        if (bobRow == null) {
            fail(name, "expected a row for Bob from the LEFT JOIN")
            return
        }
        if (bobRow.total != null) {
            fail(name, "expected null total for a user with no orders, got ${bobRow.total}")
            return
        }
        if (rows.none { it.id == createdUserId && it.total != null }) {
            fail(name, "expected Alice's row to carry a non-null total")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testCountUsersByStatus() {
    val name = "CountUsersByStatus"
    try {
        val result = countUsersByStatus(UserStatus.ACTIVE)
        if (result.status != UserStatus.ACTIVE) {
            fail(name, "expected status ACTIVE, got ${result.status}")
            return
        }
        if (result.user_count < 2) {
            fail(name, "expected count >= 2, got ${result.user_count}")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testGetUserWithTags() {
    val name = "GetUserWithTags"
    try {
        transaction {
            exec("INSERT INTO tags (name) VALUES ('vip')")
            exec("INSERT INTO user_tags (user_id, tag_id) SELECT $createdUserId, id FROM tags WHERE name = 'vip'")
        }
        val rows = getUserWithTags(createdUserId)
        if (rows.size != 1) {
            fail(name, "expected 1 tagged row, got ${rows.size}")
            return
        }
        if (rows[0].tag_name != "vip") {
            fail(name, "expected tag_name vip, got ${rows[0].tag_name}")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testSearchUsers() {
    val name = "SearchUsers"
    try {
        val users = searchUsers("Ali%")
        if (users.none { it.name == "Alice" }) {
            fail(name, "expected Alice among search results")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

// Board #197/#204: a nullable enum and a nullable composite column, each read
// both present and as SQL NULL, plus a composite field holding a double quote
// and a comma so the composite text parser's escaping is exercised rather
// than assumed. Seeded through raw exec because a composite literal is
// outside this backend's parameter-binding surface; the read path is what is
// under test and it runs entirely through generated code.
fun testGetUserProfileNullable() {
    val name = "GetUserProfile"
    try {
        var presentId = 0
        var absentId = 0
        var quotedId = 0
        transaction {
            exec(
                "INSERT INTO users (name, email, status, secondary_status, address) VALUES " +
                    "('Carol', 'carol@example.com', 'active', 'inactive', " +
                    "ROW('1 Main St', 'Springfield', '12345')) RETURNING id",
                explicitStatementType = StatementType.SELECT
            ) { rs -> if (rs.next()) presentId = rs.getInt(1); presentId }
            exec(
                "INSERT INTO users (name, email, status, secondary_status, address) VALUES " +
                    "('Dave', 'dave@example.com', 'active', NULL, NULL) RETURNING id",
                explicitStatementType = StatementType.SELECT
            ) { rs -> if (rs.next()) absentId = rs.getInt(1); absentId }
            exec(
                "INSERT INTO users (name, email, status, secondary_status, address) VALUES " +
                    "('Eve', 'eve@example.com', 'active', 'banned', " +
                    "ROW('12 \"Main\", Apt 3', 'Springfield', '12345')) RETURNING id",
                explicitStatementType = StatementType.SELECT
            ) { rs -> if (rs.next()) quotedId = rs.getInt(1); quotedId }
        }

        val present = getUserProfile(presentId)
        if (present.secondary_status != UserStatus.INACTIVE) {
            fail(name, "expected secondary_status INACTIVE, got ${present.secondary_status}")
            return
        }
        val address = present.address
        if (address == null) {
            fail(name, "expected address to be present")
            return
        }
        if (address.street != "1 Main St" || address.city != "Springfield" || address.zip != "12345") {
            fail(name, "unexpected address: $address")
            return
        }

        val absent = getUserProfile(absentId)
        if (absent.secondary_status != null) {
            fail(name, "expected null secondary_status, got ${absent.secondary_status}")
            return
        }
        if (absent.address != null) {
            fail(name, "expected null address, got ${absent.address}")
            return
        }

        val quoted = getUserProfile(quotedId)
        val quotedAddress = quoted.address
        if (quotedAddress == null) {
            fail(name, "expected an address for the quoted row")
            return
        }
        if (quotedAddress.street != "12 \"Main\", Apt 3") {
            fail(name, "composite escaping lost the quotes/comma: got ${quotedAddress.street}")
            return
        }

        transaction {
            exec("DELETE FROM users WHERE id IN ($presentId, $absentId, $quotedId)")
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testDeleteOrdersByUser() {
    val name = "DeleteOrdersByUser"
    try {
        val count = deleteOrdersByUser(createdUserId)
        if (count != 1) {
            fail(name, "expected 1 deleted order, got $count")
            return
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}

fun testDeleteUser() {
    val name = "DeleteUser"
    try {
        transaction {
            exec("DELETE FROM user_tags WHERE user_id = $createdUserId")
        }
        deleteUser(createdUserId)
        deleteUser(bobId)
        try {
            getUserById(createdUserId)
            fail(name, "expected getUserById to throw after deletion, but it returned a row")
            return
        } catch (expected: NoSuchElementException) {
            // expected: the user was deleted
        }
        pass(name)
    } catch (e: Exception) {
        fail(name, e)
    }
}
